src/main/java/com/google/cloud/spanner/pgadapter/statements/CopyStatement.java

```diff
@@ -0,0 +1,267 @@
+
+  /** @return List of column names specified in COPY statement, if provided. */
+  public List<String> getCopyColumnNames() {
+    return this.copyColumnNames;
+  }
+
+  protected void setCopyColumnNames(List<String> copyColumnNames) {
+    this.copyColumnNames = copyColumnNames;
+  }
+
+  /** @return Format type specified in COPY statement, if provided. */
+  public String getFormatType() {
+    return this.formatType;
+  }
+
+  protected void setFormatType(String formatType) {
+    this.formatType = formatType;
+  }
+
+  /** @return Null string specified in COPY statement, if provided. */
+  public String getNullString() {
+    return this.nullString;
+  }
+
+  protected void setNullString(String nullString) {
+    this.nullString = nullString;
+  }
+
+  /** @return Delimiter character specified in COPY statement, if provided. */
+  public char getDelimiterChar() {
+    return this.delimiterChar;
+  }
+
+  protected void setDelimiterChar(char delimiterChar) {
+    this.delimiterChar = delimiterChar;
+  }
+
+  /** @return Escape character specified in COPY statement, if provided. */
+  public char getEscapeChar() {
+    return this.escapeChar;
+  }
+
+  protected void setEscapeChar(char escapeChar) {
+    this.escapeChar = escapeChar;
+  }
+
+  /** @return Quote character specified in COPY statement, if provided. */
+  public char getQuoteChar() {
+    return this.quoteChar;
+  }
+
+  protected void setQuoteChar(char quoteChar) {
+    this.quoteChar = quoteChar;
+  }
+
+  /** @return True if copy data contains a header, false otherwise. */
+  public boolean hasHeader() {
+    return this.hasHeader;
+  }
+
+  protected void setHasHeader(boolean hasHeader) {
+    this.hasHeader = hasHeader;
+  }
+
+  public int getFormatCode() {
+    return this.formatCode;
+  }
+
+  protected void setFormatCode(int columnCount) {
+    this.formatCode = columnCount;
+  }
+
+  public MutationBuilder getMutationBuilder() {
+    return this.mutationBuilder;
+  }
+
+  protected void setMutationBuilder(MutationBuilder mutationBuilder) {
+    this.mutationBuilder = mutationBuilder;
+  }
+
+  private void verifyCopyColumns() throws SQLException {
+    if (this.copyColumnNames.size() > this.tableColumns.size()) {
+      throw new SQLException("Number of copy columns provided exceed table column count");
+    }
+    LinkedHashMap<String, TypeCode> tempTableColumns = new LinkedHashMap<>();
+    // Verify that every copy column given is a valid table column name
+    for (String copyColumn : this.copyColumnNames) {
+      if (!this.tableColumns.containsKey(copyColumn)) {
+        throw new SQLException(
+            "Column \"" + copyColumn + "\" of relation \"" + this.tableName + "\" does not exist");
+      }
+      // Update table column ordering with the copy column ordering
+      tempTableColumns.put(copyColumn, tableColumns.get(copyColumn));
+    }
+    this.tableColumns = tempTableColumns;
+  }
+
+  private static TypeCode parseSpannerDataType(String columnType) {
```

Comment:
This does not seem to support `NUMERIC`. Is that intentional? (It also does not support `JSON`, but as far as I know, that is also not supported on Spangres).

Would it maybe be better to change the implementation of this so that it:
1. Removes any optional length specification at the end of the type name (i.e. automatically strip `(100)` from `STRING(100)`.
2. Look up the `TypeCode` directly by looking for an enum constant with the same name as the type name without length?

If there are specific types that we do not want to support, then we could add specific code for that after the above mentioned lookup.

That would make this easier to maintain in case of new types.

Reply:
Done.